### DISCUSSION TITLE
Templ Component inside another component? Is it possible?

### DISCUSSION THREAD

**Discussion by Blankeos** [a-h/templ#554] [Q&A] [answered]
I'm learning templ, go, and htmx currently. Not a lot of resources out there so I'm struggling a lot. Is it possible to do something like this?

```go
package web

templ CountResult(newCount string) {
	<input id="" name="count" value={ newCount }/>
	<span>New Count: { newCount }</span>
	<button type="submit">
		Increment
	</button>
}

templ HelloForm(count string) {
	@Base() {
		<form hx-post="/hello" method="POST" hx-target="#hello-container">
			<input id="name" name="name" type="text"/>
			<button type="submit">Submit</button>
		</form>
		<div id="hello-container"></div>
		<form hx-post="/increment" method="POST">
		{ CountResult(count) }
		</form>
	}
}

templ HelloPost(name string) {
	<div>Hello, { name }</div>
}

```

**Comment by Blankeos** [comment, marked as answer]
Wait! I didn't read the docs enough. Lol, it was just this simple. Found it here: https://templ.guide/syntax-and-usage/template-composition :
```go
package web

templ CountResult(newCount string) {
	<input id="" name="count" value={ newCount }/>
	<span>New Count: { newCount }</span>
	<button type="submit">
		Increment
	</button>
}

templ HelloForm(count string) {
	@Base() {
		<form hx-post="/hello" method="POST" hx-target="#hello-container">
			<input id="name" name="name" type="text"/>
			<button type="submit">Submit</button>
		</form>
		<div id="hello-container"></div>
		<form hx-post="/increment" method="POST">
			@CountResult(count)
		</form>
	}
}

templ HelloPost(name string) {
	<div>Hello, { name }</div>
}
```

I'm dumm!!!! Anyway...

**Reply by a-h** [reply]
Glad the docs were useful!

**Comment by gabrielvincent** [comment]
Is it possible to have components in different files and still be able to nest them?

This is my project structure:

```
project/
├── components/
│   ├── home/
│   │   ├── home.templ
│   ├── components.templ
```

And this is the code:

```go
// home.templ
package components

templ Home() {
	@Layout() {
		<h1>
			Hello world!
		</h1>
	}
}


// components.templ
package components

templ Layout() {
	<html>
		<head>
			<title>My Project</title>
		</head>
		<body>
			<main>
				{ children... }
			</main>
		</body>
	</html>
}
```

This causes Layout to not be found. `templ generate` runs fine and generates all the files in their locations, but I get a compilation error:

```
components/home/home_templ.go:41:24: undefined: Layout
```

If I move `Home` to `components.templ` and run `templ generate` again, I no longer see the compilation error. Is this currently a limitation? Nested components must be declared in the same file?

**Reply by a-h** [reply]
No, it's just that in Go, folders correspond to package names, so given your structure:

```
project/
├── components/
│   ├── home/
│   │   ├── home.templ
│   ├── components.templ
```

Your code would need to import the parent package:

```html
// home.templ
package home
// This ^ has changed

import "github.com/gabrielvincent/project/components"
// This ^ has been added

templ Home() {
	@components.Layout() {
         // This ^ has changed
		<h1>
			Hello world!
		</h1>
	}
}
```

**Reply by gabrielvincent** [reply]
Oh, I see! This helped a lot. Thank you!

**Reply by trileminh94** [reply]
I think this point should be added into the docs, I almost quit using templ unless I found this answer. Thank you! @a-h

**Reply by a-h** [reply]
@trileminh94 - good idea. Where do you think it should be added?

I'd take a PR for it if you have time. The docs are just markdown in the ./docs directory.

**Reply by gabrielvincent** [reply]
If @trileminh94 doesn't have the time, I can create a PR adding this to the docs. I think it would make sense to include it in [Syntax and Usage > Template Composition](https://github.com/a-h/templ/blob/main/docs/docs/03-syntax-and-usage/09-template-composition.md).
